vault/quotas/quotas_rate_limit.go

```diff
@@ -181,24 +256,59 @@ func (rlq *RateLimitQuota) allow(req *Request) (Response, error) {
 		return resp, fmt.Errorf("missing request client address in quota request")
 	}
 
-	limit, remaining, reset, allow := rlq.store.Take(req.ClientAddress)
-	resp.Allowed = allow
-	resp.Headers = map[string]string{
-		httplimit.HeaderRateLimitLimit:     strconv.FormatUint(limit, 10),
-		httplimit.HeaderRateLimitRemaining: strconv.FormatUint(remaining, 10),
-		httplimit.HeaderRateLimitReset:     time.Unix(0, int64(reset)).UTC().Format(time.RFC1123),
+	var retryAfter string
+
+	defer func() {
+		if !resp.Allowed {
+			resp.Headers[httplimit.HeaderRetryAfter] = retryAfter
+			rlq.metricSink.IncrCounterWithLabels([]string{"quota", "rate_limit", "violation"}, 1, []metrics.Label{{"name", rlq.Name}})
+		}
+	}()
+
+	rlq.lock.Lock()
+	defer rlq.lock.Unlock()
+
+	// Check if the client is currently blocked and if so, deny the request. Note,
+	// we cannot simply rely on the presence of the client in the map as the timing
+	// of purging blocked clients may not yield a false negative. In other words,
+	// a client may no longer be considered blocked whereas the purging interval
+	// has yet to run.
+	if blockedAt, ok := rlq.blockedClients[req.ClientAddress]; ok {
```

Comment:
Using a sync.Map would mean not every request was contending for the same lock here, too.  We would only need rlq.lock further down, to check after we were disallowed.

(I think that's the case anyway-- memorystore.Store does its own locking, correct? And we don't need the mutex to access rlq.store?)

Reply:
Correct, we don't need a lock for the `store`, the rate limiter is already thread-safe. See response in the first comment.